### PR TITLE
feat: allow empty pattern match

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes https://github.com/flix/flix/issues/6007

### CODE REVIEW

**Comment by magnus-madsen** [comment]
I see you started making changes to the AST, but really the goal is simply to keep the same AST and just support an empty set of cases.

**Comment by jaschdoc** [comment]
Roger

File: main/test/flix/Test.Exp.Match.Empty.flix [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+mod Test.Exp.Match.Empty {
+
+    @test
+    def testEmptyMatch01(): Bool = match (???: Void) { }
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.Match.Empty.flix:20]: We cannot test this feature really :)

Or rather, only by checking that code compiles, not that it runs :)

File: main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala [hunk 2]
```diff
@@ -1851,4 +1851,40 @@ class TestTyper extends AnyFunSuite with TestUtils {
...
+        |def f(): String = match (???: String) { }
+        |""".stripMargin
+    val result = compile(input, Options.TestWithLibNix)
+    expectSuccess(result)
+  }
+
+  test("TestMatchEmpty.03") {
+    val input =
+      """
+        |def f(): String = match (42) { }
```
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1876]: Didn't you say we should allow `match e { }` where `e` is an expression right?

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1876]: Allow `match e {}` but rejected it in PatMatch if `e` is not the empty type?

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1876]: Like the point is that: All pattern matches should be exhaustive :) As today

It just so happens that if `e` has AN EMPTY TYPE then the EMPTY list of patterns is exhaustive :)

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala:1876]: It is okay for me do nothing if you give me all your potatoes and tell me to peel them _if I know you have zero potatoes_ :)

**Comment by magnus-madsen** [comment]
I think the real problem here is that I never explained _WHAT_ I was trying to achieve. So that is a bit on me. Let me try:

**Comment by magnus-madsen** [comment]
We want to add algebraic effects, a special case is exceptions:

```
eff MyException {
  pub def throwIt() // What return type should go here? Intuitively a throw can never return, i.e. there is NO value. 
  // We can handle this by "returning" Void
}
```

where

```
enum Void {}
```

Now lets try to use the effect:

```
def div(x: Int32, y: Int32): Int32 \ {MyThrow} = 
  if (y == 0) do MyThrow.throwIt() else x / y
```

Woops. This does not type check! Why? Because one branch returns `Void` and the other branch returns `Int32` (!!!) (Recall if-then-else requires the same type of both branches). Whats the solution? If there is subtyping one could make `Void` a subtype of everything, but we don't have that. Instead, pattern match on `Void`:

```
def div(x: Int32, y: Int32): Int32 \ {MyThrow} = 
  if (y == 0) match (do MyThrow.throwIt()) {} else x / y
```

the match has type `a` which unifies with `Int32` so now everything works.

This is of course a bit ugly, so introduce `absurd e = match e {}` and write:

```
def div(x: Int32, y: Int32): Int32 \ {MyThrow} = 
  if (y == 0) absurd (do MyThrow.throwIt()) else x / y
```

**Comment by magnus-madsen** [comment]
So that is what we are trying to achieve. If-- for some reason-- I have `e: Void` then I want to be able to write `match e {}`, have it type check, let it have a polymorphic type `a`, and let it pass the patmatch phase.

**Comment by jaschdoc** [comment]
Why do we have to make such a big change, then? We could easily just add syntactic sugar for:
```scala
pub def f(): String = match ??? { case _ => ??? }
```
This compiles and throws a runtime error.

**Comment by magnus-madsen** [comment]
Not sure I follow :) The point is we need `Void` to support AE that emulate exceptions.

**Comment by magnus-madsen** [comment]
How does the PatMatch behave if left unchanged when matching on Void? Does it crash?

**Comment by jaschdoc** [comment]
I'm not sure, I understand the importance of the `Void` type specifically.

The match expression below is polymorphic so if we add the `Void` type later for AE, then the below expression should unify with `Void`.
```scala
pub def f(): String = match ??? { case _ => ??? }
```

You can try it out currently with any type and it works but obviously crashes. My suggestion is that we just desugar `absurd e` to
```
match e { case _ => ?absurd }
```

**Comment by magnus-madsen** [comment]
I will dwell a bit more on this "feature".

**Pull request closed without merging**